slackclient/slackrequest.py

```diff
@@ -94,12 +104,89 @@ def do(self, token, request="?", post_data=None, domain="slack.com", timeout=Non
             if isinstance(v, (list, dict)):
                 post_data[k] = json.dumps(v)
 
+        request_args = {
+            'api_method': api_method,
+            'post_data': post_data,
+        }
+        return self.__post_http_request(token, request_args, files, timeout)
+
+    def __post_http_request(self, token, request_args, files=None, timeout=None):
+        # Check the access token's expiration timestamp before submitting
+        # the API request and refresh if expired
+        if(self.refresh_token and self.access_token_expires_at):
+            current_ts = int(datetime.now().strftime("%s")) * 1000
+            if(current_ts > self.access_token_expires_at):
+                self.access_token_expires_at = None
+                self.refresh_access_token()
+
+        # This allows us to override the default token (self.access_token) for refresh requests
+        if(self.access_token and token is None):
+            token = self.access_token
+
+        # Set user-agent and auth headers
+        headers = {
+            'user-agent': self.get_user_agent(),
+            'Authorization': 'Bearer {}'.format(token)
+        }
+
         # Submit the request
-        return requests.post(
-            url,
+        res = requests.post(
+            'https://{0}/api/{1}'.format(self.domain, request_args['api_method']),
             headers=headers,
-            data=post_data,
+            data=request_args['post_data'],
             files=files,
             timeout=timeout,
             proxies=self.proxies
         )
+        response_json = res.json()
+        # if the API request returns an invalid_auth error, refresh the token and try again
+        if (res.status_code is 200 and response_json['ok'] is False):
```

Comment:
i think the post-processing of the response significantly alters the behavior of `SlackRequest`. this can be moved further up the stack to either `Server` or `SlackClient` (preferably, since the `Server` class is a messy abstraction we plan to get rid of in the future).

this would also mean that much of the data being captured in the `SlackRequest` instance (by the constructor) is no longer necessary, and that would eliminate a class of errors that could stem from that data not remaining in sync across all these objects.

Reply:
Moved this logic into `SlackClient`.